### PR TITLE
fix issue 3031 - static var and func declaration conflict in different sub scope

### PR DESCRIPTION
**Pull request by ghost**
I'm glad to propose a fix for this very old (in IT time, 11 years is much !) issue.

The approach of the fix is based on the observation that the fundamental (and only) problem was that allowing such declarations would produce same mangling.
Now when the insertion in the scope fails, and if the declaration was static, its name is changed and a **local** alias declaration, taking the previous name, is generated.

concrectly, some code showing the lowering:

```d
void main()
{
    // the following ...
    { static int a; }
    { static int a; } // same mangling, but mangling is not useful !
    // is rewritten as
    { static int a; }
    { static int newID; alias a = newID; }
}
```

This works because:
- the static declaration, the one that needs a unique mangle, has now an unique identifier.
- the code following the declaration use the new alias.

### CODE REVIEW

**Comment by Geod24** [comment]
I think this is also related to the aggregated bug Kenji created: https://issues.dlang.org/show_bug.cgi?id=14831
You might want to look it up to see which issues he ran into, particularly the `extern (C[++])` ones.

File: src/dmd/dscope.d [hunk 1]
```diff
@@ -66,6 +66,8 @@ enum SCOPE
     // The following are mutually exclusive
     printf        = 0x4_0000, /// printf-style function
     scanf         = 0x8_0000, /// scanf-style function
+
+    subfuncscope  = 0x100_000, /// in a function and in a compound scope
```
Comment (ibuclaw) [inline src/dmd/dscope.d:70]: It might seem counterintuitive, but it would be good to have the underscore in the same position as the values above (`0x10_0000`).

File: src/dmd/expressionsem.d [hunk 2]
```diff
@@ -5129,12 +5129,40 @@ private extern (C++) final class ExpressionSemanticVisitor : Visitor
...
+                    s.ident             = Identifier.generateId("issue3031_");
+                    AliasDeclaration ad = new AliasDeclaration(s.loc, oldID, s);
+
+                    dsymbolSemantic(s, sc);
+                    dsymbolSemantic(ad, sc);
+
+                    sc.func.localsymtab.insert(ad);
+                    if (!sc.func.localsymtab.insert(s))
+                    {
+                        e.error("fail to apply fix for bugzilla 3031");
```
Comment (ibuclaw) [inline src/dmd/expressionsem.d:5158]: Error message lacks meaning or reason, you might as well say `compiler inhibited`.

File: src/dmd/expressionsem.d [hunk 3]
```diff
@@ -5129,12 +5129,40 @@ private extern (C++) final class ExpressionSemanticVisitor : Visitor
...
+                //  the fix is to replace the static decl name and use the old sym name as **local** alias,
+                //  to get            `{ { static T decl; } { static T RndID; alias decl = RndID; }`
+                //  as 1) it is not possible to link against a global var that is hidden in a func body.
+                //     2) in a func body, out of order decls are not supported so we can patch the name without affecting prior code
+                if (!inserted && (sc.flags & SCOPE.subfuncscope) &&
+                    ((v && v.isDataseg()) || ed || (fd && fd.storage_class & STC.static_)))
+                {
+                    // printf("fixing 3031, on %s at line %d\n", s.toChars(), s.loc.linnum);
+                    Identifier oldID    = s.ident;
+                    s.ident             = Identifier.generateId("issue3031_");
```
Comment (ibuclaw) [inline src/dmd/expressionsem.d:5149]: This name is unrelated to the original, how on earth are you supposed to debug print nested static symbols?

Comment (NilsLankila) [reply src/dmd/expressionsem.d:5149]: true in theory but in practice debugging local statics does not work
thory_vs_hjkl

Comment (MoonlightSentinel) [reply src/dmd/expressionsem.d:5149]: Maybe rename those variables as `<original name>_<line>_<column>`?

Comment (NilsLankila) [reply src/dmd/expressionsem.d:5149]: yes good idea.

Comment (rainers) [reply src/dmd/expressionsem.d]: On Windows, debug info for statics exists and can be used in watches, but they are not enumerated with the locals (I guess it is the same for ELF object files). LDC/GDC might be different, too.

The `<original name>_<line>_<column>` pattern isn't good enough for templates, I think your test case should expose this with a third instantiation.

An alternative might be to keep the identifier as is, but to insert a parent symbol for the scope in the mangled name.

Comment (ibuclaw) [reply src/dmd/expressionsem.d]: > true in theory but in practice debugging local statics does not work

It has worked in gdb in the past.

File: src/dmd/dscope.d [hunk 4]
```diff
@@ -66,6 +66,8 @@ enum SCOPE
     // The following are mutually exclusive
     printf        = 0x4_0000, /// printf-style function
     scanf         = 0x8_0000, /// scanf-style function
+
+    subfuncscope  = 0x10_0000, /// in a function and in a compound scope
```
Comment (WalterBright) [inline src/dmd/dscope.d:70]: Is this really needed? The old code was able to unambiguously detect these cases without it.

File: src/dmd/expressionsem.d [hunk 5]
```diff
@@ -5129,12 +5129,40 @@ private extern (C++) final class ExpressionSemanticVisitor : Visitor
...
+                //  the fix is to replace the static decl name and use the old sym name as **local** alias,
+                //  to get            `{ { static T decl; } { static T RndID; alias decl = RndID; }`
+                //  as 1) it is not possible to link against a global var that is hidden in a func body.
+                //     2) in a func body, out of order decls are not supported so we can patch the name without affecting prior code
+                if (!inserted && (sc.flags & SCOPE.subfuncscope) &&
+                    ((v && v.isDataseg()) || ed || (fd && fd.storage_class & STC.static_)))
+                {
+                    // printf("fixing 3031, on %s at line %d\n", s.toChars(), s.loc.linnum);
+                    scope OutBuffer ob;
+                    ob.printf("%s_L%d_C%d_id", s.ident.toChars(), s.loc.linnum, s.loc.charnum);
```
Comment (WalterBright) [inline src/dmd/expressionsem.d:5149]: Compiler generated symbols start with `__` so they won't conflict with user names. I suggest instead:
```
"__id_%s_%d_%d"
```

File: src/dmd/expressionsem.d [hunk 6]
```diff
@@ -5129,12 +5129,40 @@ private extern (C++) final class ExpressionSemanticVisitor : Visitor
...
+
+                    inserted = true;
+                }
                 // https://issues.dlang.org/show_bug.cgi?id=11720
                 // include Dataseg variables
-                if ((s.isFuncDeclaration() ||
-                     s.isAggregateDeclaration() ||
-                     s.isEnumDeclaration() ||
-                     v && v.isDataseg()) && !sc.func.localsymtab.insert(s))
+                if (!inserted && (fd || ag || ed || (v && v.isDataseg())))
```
Comment (WalterBright) [inline src/dmd/expressionsem.d:5165]: `inserted` can be dispensed with if `if` is replaced with `else if`.

**Comment by WalterBright** [comment]
My idea was to fix this in the name mangling code. I like your idea of using an alias better.

File: src/dmd/expressionsem.d [hunk 7]
```diff
@@ -5129,12 +5129,40 @@ private extern (C++) final class ExpressionSemanticVisitor : Visitor
             }
             else if (sc.func)
             {
+                FuncDeclaration fd      = s.isFuncDeclaration();
+                AggregateDeclaration ag = s.isAggregateDeclaration();
+                EnumDeclaration ed      = s.isEnumDeclaration();
+                bool inserted     =  (fd || ag || ed || (v && v.isDataseg())) && sc.func.localsymtab.insert(s);
```
Comment (WalterBright) [inline src/dmd/expressionsem.d:5135]: This is a hot path, and the expression is computed multiple times. Replace with:
```
const bool isFAE = (fd = s.isFuncDeclaration()) !is null ||
                               s.isAggregateDeclaration() !is null ||
                               (ed = s.isEnumDeclaration()) !is null ||
                               (v && v.isDataseg());
```
and ag will never even be used.

File: src/dmd/expressionsem.d (hunk 6 above)
Comment (NilsLankila) [reply src/dmd/expressionsem.d:5165]: it is required because otherwise after a successful insertion  (e.g when there's not conflict) the path of the `else if` branch is taken.

File: src/dmd/dscope.d (hunk 4 above)
Comment (NilsLankila) [reply src/dmd/dscope.d]: it seems it wasnt.

File: src/dmd/expressionsem.d (hunk 7 above)
Comment (NilsLankila) [reply src/dmd/expressionsem.d]: ag is now used, this was not test before the last force push

File: src/dmd/expressionsem.d [hunk 8]
```diff
@@ -5130,12 +5130,40 @@ private extern (C++) final class ExpressionSemanticVisitor : Visitor
...
+                //  second `decl` insertion fails.
+                //  the fix is to replace the static decl name and use the old sym name as **local** alias,
+                //  to get            `{ { static T decl; } { static T RndID; alias decl = RndID; }`
+                //  as 1) it is not possible to link against a global var that is hidden in a func body.
+                //     2) in a func body, out of order decls are not supported so we can patch the name without affecting prior code
+                bool insertedFAEV = isFAEV ? sc.func.localsymtab.insert(s) !is null : true;
+                if (!insertedFAEV)
+                {
+                    // printf("fixing 3031, on %s at line %d\n", s.toChars(), s.loc.linnum);
+                    scope OutBuffer ob;
```
Comment (ibuclaw) [inline src/dmd/expressionsem.d:5150]: Scope is redundant here, OutBuffer is already a struct.

File: test/compilable/test1537.d [hunk 9]
```diff
@@ -25,44 +25,6 @@ static assert( __traits(identifier, bug4732) == "bug4732");
...
-
-int bug5245d(U)()
-{
-    { enum  T { a = 3 } T w; }
-    { struct T { int a = 2; } T v; }
-    return 91;
-}
-
-
-static assert(!is(typeof(Compileable!(bug5245a!(int)()).OK)));
```
Comment (ibuclaw) [inline test/compilable/test1537.d:59]: It would be good to keep these tests around with a comment saying as of [bugzilla reference], test is now compilable.

Comment (NilsLankila) [reply test/compilable/test1537.d:59]: yeah but I dont know. I don't even know why they were here. They looked like a primitive way to do something like `__traits(compiles)`.

Comment (ibuclaw) [reply test/compilable/test1537.d:59]: Yes, this predates `__traits`, but don't adapt it, as it is still a code path that should remain tested.

Comment (NilsLankila) [reply test/compilable/test1537.d:59]: to be clear you want me to change `!is` into `is`, right ?

Comment (NilsLankila) [reply test/compilable/test1537.d:59]: okay done

Comment (ibuclaw) [reply test/compilable/test1537.d:59]: Yes.

**Comment by NilsLankila** [comment]
Is this good now ?

File: test/compilable/fail11720.d [hunk 10]
```diff
@@ -0,0 +1,23 @@
+void main()
+{
+    foo();
```
Comment (MoonlightSentinel) [inline test/compilable/fail11720.d:3]: This test should create multiple different instantiations of `foo` given https://github.com/dlang/dmd/pull/11250#discussion_r437891414

Comment (NilsLankila) [reply test/compilable/fail11720.d:3]: I don't follow you, it's only called once.

Comment (NilsLankila) [reply test/compilable/fail11720.d:3]: can you expand the problem ?

Comment (rainers) [reply test/compilable/fail11720.d:3]: The problem I meant is that there are only 2 statics generated in foo, one generating the original symbol, one with the aliased symbol. A third static would verify that the generated symbols don't collide. IIRC the version I commented on didn't use `generateId` so it might be solved now to some extend, but a test (a third static generated at the same location) could verify that.

There is still a problem with `generateId`, though: with separate compilation units, the same symbol name can be generated, but might have different implementations.

Comment (NilsLankila) [reply test/compilable/fail11720.d:3]: Ok thanks for the explanations. Actually I'm embarassed because this was not supposed to be covered by the PR but it had appeared that the test passed.

**Comment by MoonlightSentinel** [comment]
Another solution was implemented
